Resumable uploads

pdecat opened the pull request:
That PR implements resumable uploads.

For now, I've implemented the upload url storage in a separate database to avoid polluting the main database while testing. That decision can be revised if desired.

Note: this depends on https://github.com/gphotosuploader/google-photos-api-client-go/pull/5

pdecat commented:
Hint to kill an upload in progress to verify the proper behavior:

```
# sudo tcpkill -i eno1 host $(netstat -anp|grep 443|grep gphotos|awk -F "[ :]+" '{print $6}')
```

pdecat commented:
Rebased on master following merges of #45 and #67.

pdecat commented:
Rebased on master following merge of #68.

pdecat commented:
Rebased on master to fix conflicts.

pacoorozco commented on go.mod:
```diff
@@ -23,3 +22,7 @@ require (
 	gopkg.in/mgo.v2 v2.0.0-20180705113604-9856a29383ce // indirect
 )
+
+replace github.com/gphotosuploader/google-photos-api-client-go => github.com/pdecat/google-photos-api-client-go v0.0.0-20190728103106-52b4c499bcab
```
Why are you using this replace? We can merge your code into `github.com/gphotosuploader/google-photos-api-client-go`

pdecat replied:
This replace is needed to make it build here until the google-photos-api-client-go PR is merged.

pacoorozco commented:
Hi @pdecat 

Once https://github.com/gphotosuploader/google-photos-api-client-go/pull/5 has been merged. Can you rebase and solve conflicts in order to be able to merge this PR?

Can we drop **[WIP]** from title?

Thanks

pdecat commented:
Rebased on master and removed WIP from title.

Commits related to refresh token were moved to https://github.com/gphotosuploader/gphotos-uploader-cli/pull/107

pdecat commented:
Rebased on master.

pdecat commented:
Squashed commits.

pacoorozco commented on config/config.go:
```diff
@@ -42,11 +42,16 @@ func NewConfig(dir string) *Config {
 
+// UploadURLsTrackingDBDir returns the path of the folder where upload URLs are tracked.
+func (c *Config) UploadURLsTrackingDBDir() string {
+	return path.Join(c.ConfigPath, "uploadsurls.db")
```
I don't like to have `uploads.db` and `uploadsurls.db`. They are very similar, but changing the first one would be a breaking change and I can't figure out a better name for the second one... maybe `resumable_uploads.db`

pdecat replied:
Agreed, done.

pacoorozco commented on config/config_test.go:
```diff
@@ -94,7 +94,7 @@ func TestConfig_TrackingDBDir(t *testing.T) {
 
 	expected := path.Join(dir, "uploads.db")
-	got := c.TrackingDBDir()
+	got := c.CompletedUploadsTrackingDBDir()
```
Please add a test for the new resumable uploads DB dir, something similar to:

```
func TestConfig_ResumableUploadsDBDir(t *testing.T) {
	dir := filepath.Join(os.TempDir(), fmt.Sprintf("gphotos-config.%d", time.Now().UnixNano()))
	c := config.NewConfig(dir)

	expected := path.Join(dir, "resumable_uploads.db")
	got := c.ResumableUploadsDBDir()

	if got != expected {
		t.Errorf("Testing get resumable uploads DB dir: expected: %s, got %s", expected, got)
	}

}
```

pdecat replied:
Done.

pacoorozco commented on upload/fileUpload.go:
```diff
@@ -71,14 +73,37 @@ func getGooglePhotosAlbumID(name string, c *gphotos.Client) string {
+	curUploadURL, err := uploadURLsService.GetUploadURL(f.path)
+	if err != nil {
+		// Not found, not an error, just an empty upload URL
+		log.Println(err)
```
I will add here `curUploadURL = ""`. I will not trust on the response of `uploadURLsService.GetUploadURL)` in case of error.

pdecat replied:
`GetUploadURL()` returns `""` in all error cases, but as you will.

Done.

pacoorozco replied:
It returns now... but eventually could return another thing... when a function gives an error, we should not trust on the returned value. :+1

pacoorozco commented on upload/fileUpload.go:
```diff
@@ -71,14 +73,37 @@ func getGooglePhotosAlbumID(name string, c *gphotos.Client) string {
 	// upload the file content to Google Photos
-	_, err := f.client.UploadFile(f.path, albumID)
+	ptrUploadURL := &curUploadURL
+	_, err = f.client.UploadFileResumable(f.path, ptrUploadURL, albumID)
```
TODO: We can create a `ResumableFile` object that has a path and a url...

pacoorozco commented on upload/fileUpload.go:
```diff
@@ -71,14 +73,37 @@ func getGooglePhotosAlbumID(name string, c *gphotos.Client) string {
+	}
+
+	if err != nil && *ptrUploadURL != "" {
+		log.Printf("Error uploading file '%s', storing upload URL '%s'\n", f.path, *ptrUploadURL)
```
You don't need `\n` ending. `log` put it as default.

pdecat replied:
Indeed, done.

pdecat commented on cmd/root.go:
```diff
@@ -56,9 +57,9 @@ func startUploader(cmd *cobra.Command, args []string) {
 
 	// load completedUploads DB
-	db, err := leveldb.OpenFile(cfg.TrackingDBDir(), nil)
+	db, err := leveldb.OpenFile(cfg.CompletedUploadsTrackingDBDir(), nil)
```
Done.

pdecat commented on cmd/root.go:
```diff
@@ -74,8 +75,16 @@ func startUploader(cmd *cobra.Command, args []string) {
+	// load upload URLs DB
+	uploadURLsdb, err := leveldb.OpenFile(cfg.UploadURLsTrackingDBDir(), nil)
+	if err != nil {
+		log.Fatalf("Error opening upload URLs db: path=%s, err=%v", cfg.UploadURLsTrackingDBDir(), err)
```
Done.

pdecat commented on cmd/root.go:
```diff
@@ -74,8 +75,16 @@ func startUploader(cmd *cobra.Command, args []string) {
 	tkm := tokenstore.NewService(kr)
 
+	// load upload URLs DB
+	uploadURLsdb, err := leveldb.OpenFile(cfg.UploadURLsTrackingDBDir(), nil)
```
Done.
